src/model/sprite.ts

```diff
@@ -0,0 +1,205 @@
+/*
+This file handles operations related to Sprites: loading them and getting info about them.
+A Sprite is a sub-image of a parent image.  The sub image is represented by a rectangle
+within the parent image (the four numbers: x, y, width, height).
+*/
+/*
+global
+G_view_drawSprite
+G_model_createCanvas
+*/
+
+type Sprite = [
+  HTMLCanvasElement | HTMLImageElement,
+  number,
+  number,
+  number,
+  number
+];
+
+const createSprite = (
+  img: HTMLCanvasElement | HTMLImageElement,
+  x: number,
+  y: number,
+  w: number,
+  h: number
+): Sprite => {
+  return [img, x, y, w, h];
+};
+
+// sprites are stored inside key => value objects, where the key is the name of the sprite
+type SpriteCollection = { [key: string]: Sprite };
+
+// This type represents the postfix one can add to a sprite
+type SpriteModification = '' | '_r1' | '_r2' | '_r3' | '_f';
+const G_SPRITE_MOD_NORMAL: SpriteModification = '';
+const G_SPRITE_MOD_FLIPPED: SpriteModification = '_f';
+const G_SPRITE_MOD_ROT90: SpriteModification = '_r1';
+const G_SPRITE_MOD_ROT180: SpriteModification = '_r2';
+const G_SPRITE_MOD_ROT270: SpriteModification = '_r3';
+
+let model_sprites: SpriteCollection | null = null;
+
+// given an inputCanvas, return a new canvas rotated to the right by 90 degrees
+const createRotatedImg = (
+  inputCanvas: HTMLCanvasElement
+): HTMLCanvasElement => {
+  const [canvas, ctx, width, height] = G_model_createCanvas(
+    inputCanvas.width,
+    inputCanvas.height
+  );
+  const x = width / 2;
+  const y = height / 2;
+  ctx.translate(x, y);
+  ctx.rotate(Math.PI / 2);
+  ctx.drawImage(inputCanvas, -x, -y);
+  return canvas;
+};
+
+// given an inputCanvas, return a new canvas flipped horizontally
+const createFlippedImg = (
+  inputCanvas: HTMLCanvasElement
+): HTMLCanvasElement => {
+  const [canvas, ctx, width] = G_model_createCanvas(
+    inputCanvas.width,
+    inputCanvas.height
+  );
+  ctx.translate(width, 0);
+  ctx.scale(-1, 1);
+  ctx.drawImage(inputCanvas, 0, 0);
+  return canvas;
+};
+
+// given a Sprite, create and return a new image from the sprite
+const spriteToCanvas = (sprite: Sprite): HTMLCanvasElement => {
+  const [, , , spriteWidth, spriteHeight] = sprite;
```

Comment:
What is this witchcraft?

Reply:
A `sprite` is an array of `[image, x, y, width, height]`.  In this function, I dont need those first three values, so I can simply not declare them as variables when destructuring the array.

Reply:
Or rather, I have declared them as `nothing` and indicated to the compiler to ignore them with the `,`